File: ai-ml/modules/skill_gap_analyzer.py
from typing import List, Dict

def analyze_skill_gap(current_skills: List[str], target_skills: List[str]) -> Dict:
    current_lower = set([s.lower() for s in current_skills or []])
    target_lower = [t.lower() for t in target_skills or []]
    missing = [t for t in target_lower if t not in current_lower]
    matched = [t for t in target_lower if t in current_lower]
    return {
        'currentSkills': sorted(list(current_lower)),
        'targetSkills': target_lower,
        'matched': matched,
        'missing': missing,
        'gapSize': len(missing)
    }

File: server/utils/skillGapAnalyzer.js
function analyzeSkillGap(currentSkills = [], targetSkills = []) {
  const currentList = (currentSkills || []).map(s => String(s).toLowerCase());
  const currentSet = new Set(currentList);
  const targetList = (targetSkills || []).map(s => String(s).toLowerCase());
  const targetSet = new Set(targetList);

  const matched = targetList.filter(t => currentSet.has(t));
  const missing = targetList.filter(t => !currentSet.has(t));
  // Irrelevant: present in resume but not required by target path
  const irrelevant = currentList.filter(s => !targetSet.has(s));

  return {
    matched,
    missing,
    irrelevant,
    gapSize: missing.length,
  };
}

module.exports = { analyzeSkillGap };